pkg/kn/commands/plugin/plugin_list.go

```diff
@@ -0,0 +1,268 @@
+// limitations under the License.
+
+package plugin
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/knative/client/pkg/kn/commands"
+	"github.com/spf13/cobra"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+// ValidPluginFilenamePrefixes controls the prefix for all kn plugins
+var ValidPluginFilenamePrefixes = []string{"kn"}
+
+// NewPluginListCommand creates a new `kn plugin list` command
+func NewPluginListCommand(p *commands.KnParams) *cobra.Command {
+	pluginFlags := PluginFlags{
+		IOStreams: genericclioptions.IOStreams{
+			In:     os.Stdin,
+			Out:    os.Stdout,
+			ErrOut: os.Stderr,
+		},
+	}
+
+	pluginListCommand := &cobra.Command{
+		Use:   "list",
+		Short: "list all visible plugin executables on a user's PATH",
+		Long: `List all available plugin files on a user's PATH.
+
+		Available plugin files are those that are:
+		- executable
+		- anywhere on the user's PATH
+		- begin with "kn-`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := pluginFlags.Complete(cmd)
+			if err != nil {
+				return err
+			}
+
+			err = pluginFlags.Run()
+			if err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	pluginFlags.AddPluginFlags(pluginListCommand)
+
+	return pluginListCommand
+}
+
+func (o *PluginFlags) Complete(cmd *cobra.Command) error {
+	o.Verifier = &CommandOverrideVerifier{
+		root:        cmd.Root(),
+		seenPlugins: make(map[string]string, 0),
+	}
+
+	o.PluginPaths = filepath.SplitList(os.Getenv("PATH"))
+	return nil
+}
+
+// Run to find all kn plugins in current PATH with name matching kn-*
+func (o *PluginFlags) Run() error {
+	pluginsFound := false
+	isFirstFile := true
+	pluginErrors := []error{}
+	pluginWarnings := 0
+
+	for _, dir := range uniquePathsList(o.PluginPaths) {
+		if dir == "" {
+			continue
+		}
+		files, err := ioutil.ReadDir(dir)
+		if err != nil {
+			if _, ok := err.(*os.PathError); ok {
+				fmt.Fprintf(o.ErrOut, "Unable read directory %q from your PATH: %v. Skipping...", dir, err)
+				continue
+			}
+
+			pluginErrors = append(pluginErrors, fmt.Errorf("error: unable to read directory %q in your PATH: %v", dir, err))
+			continue
+		}
+
+		for _, f := range files {
+			if f.IsDir() {
+				continue
+			}
+			if !hasValidPrefix(f.Name(), ValidPluginFilenamePrefixes) {
+				continue
+			}
+
+			if isFirstFile {
+				fmt.Fprintf(o.ErrOut, "The following compatible plugins are available:\n\n")
+				pluginsFound = true
+				isFirstFile = false
+			}
+
+			pluginPath := f.Name()
+			if !o.NameOnly {
+				pluginPath = filepath.Join(dir, pluginPath)
+			}
+
+			fmt.Fprintf(o.Out, "%s\n", pluginPath)
+			if errs := o.Verifier.Verify(filepath.Join(dir, f.Name())); len(errs) != 0 {
+				for _, err := range errs {
+					fmt.Fprintf(o.ErrOut, "  - %s\n", err)
+					pluginWarnings++
+				}
+			}
+		}
+	}
+
+	if !pluginsFound {
+		pluginErrors = append(pluginErrors, fmt.Errorf("error: unable to find any kubectl plugins in your PATH"))
```

Comment:
No plugins shouldn't be considered an error - it's a valid state.

Reply:
Fair.